torchao/dtypes/nf4tensor.py

```diff
@@ -565,6 +845,66 @@ def __torch_function__(cls, func, types, args=(), kwargs=None):
             return func(*args, **kwargs)
 
 
+    def fsdp_pre_all_gather(self, mesh: DeviceMesh) -> Tuple[Tuple[torch.Tensor, ...], Any]:
+        return (
+            self.quantized_scalers,
+            self.quantization_factor,
+            self.quantized_data,
+        ), (
+            SubclassTensorArgs(
+                self.size(),
+                self.stride(),
+                self.storage_offset(),
+                self.dtype,
+                self.device,
+                self.requires_grad,
+            ),
+            self.block_size,
+            self.n_blocks,
+            self.scaler_block_size,
+            self.scaler_mean,
+            self.nf4,
+            mesh.get_group().size(),
```

Comment:
n00b q: what is this doing?

Reply:
> I don't follow what the 2 fsdp tests are trying to do

This is core logic in `nf4tensor.py`. unit tests happens in another file `test_nf4.py`  

`fsdp_pre_all_gather` returns a tuple of two things
* tuple[0] are `quantized_scalers`, `quantization_factor` and `quantized_data`. They are input for all-gather
* tuple[1] are `SubclassTensorArgs`, `block_size` etc are metadata to reconstruct NF4Tensor. `mesh.get_group().size()` is the group size for all-gather (how many gpus). it's helpful to restore NF4Tensor.size. Eg for 2 gpus, all-gathering tensor(512) will return tensor(512 x 2)